src/dbnode/storage/index/block.go

```diff
@@ -253,30 +257,74 @@ func (b *block) Query(
 		return false, err
 	}
 
-	var (
-		size       = results.Size()
-		brokeEarly = false
-	)
+	size := results.Size()
+	limitedResults := false
 	execCloser := safeCloser{closable: exec}
 	iterCloser := safeCloser{closable: iter}
 
+	// NB(r): iter.Next() when moving between segments will actually
+	// execute the search phase which can take a long time, hence
+	// we only retrieve the results lock when we add a batch of documents
+	// to the results set.
+	batch := b.docsPool.Get()
+	// Use documentArrayPoolCapacity to as max batch to avoid growing outside
```

Comment:
"Use the maximum capacity of the array pool as the max batch size to avoid growing outside the allowed pool capacity."